heap/heap.py

```diff
@@ -1,21 +1,121 @@
+        # 0th index of the storage with 0, we will not be using this place-holder index
+        # or the value.
+        self.storage = [0]
+        # for i in items:
+        #     self.storage.append(i)
+        #     self._bubble_up(len(self.storage)-1)
 
-  def insert(self, value):
-    pass
+    def insert(self, value):
+        # on each insert, we are going to append the value to storage, (end of list)
+        self.storage.append(value)
+        # every time we insert a value, we have to make sure the max value is up top
+        self._bubble_up(len(self.storage)-1)
 
-  def delete(self):
-    pass
+    def delete(self):
+        # to delete we have three conditions, when length of storage is 1,
+        # which means there is nothing in the heap
+        # when length is 2, which means there is one value in the heap
+        # and when it is greater than two
+        # when length of storage is greater than two, we have to swap the first index
+        # which at this point is the greatest value, with the last index in the list
+        # after which we store the max_value from the new position which is the last index
+        # in the list and then sift the new 1st index which was the last index previously
+        # if length of storage is equal to 1, return none, there is nothing to delete
+        if len(self.storage) == 1:
+            return None
+        # if the length of the storage is equal to 2, which would mean 1 value
+        # all you need to do is pop the value at the end of the list
+        if len(self.storage) == 2:
+            max_val = self.storage.pop()
+        # if the length of the storage is greater than 2 we need to swap the largest value
+        # with the last index of the list
+        if len(self.storage) > 2:
+            self._swap(1, len(self.storage)-1)
+            # after the swap we assign max_val to the last index
+            max_val = self.storage.pop()
+            # call sift down on index of one, which used to be the last index
+            self._sift_down(1)
+        # return max value
+        return max_val
 
-  def get_max(self):
-    pass
+    def get_max(self):
+        # get max checks to see if there is a value at the 1st index, if it does
+        # return that value, if not return None
+        if self.storage[1]:
+            return self.storage[1]
+        else:
+            return None
 
-  def get_size(self):
-    pass
+    def _swap(self, i, j):
+        # all we need to do in the swap function is to swap the 1st index with the last
+        self.storage[i], self.storage[j] = self.storage[j], self.storage[i]
 
-  def _bubble_up(self, index):
-    pass
+    def get_size(self):
+        # because we initialized the storage with 0 at the 0th index we must be sure
+        # to get length of storage -1
+        return len(self.storage)-1
 
-  def _sift_down(self, index):
-    pass
+    def _bubble_up(self, index):
```

Comment:
Both bubble up and sift down need to be repeated until index arrives at the right place.  A while loop might be a good choice for creating the iteration.

Reply:
i think i am repeating it, with recursion,
for the bubble up, it is going to continue to bubble up until :
        # if value at index is greater than value at parent swap
        if self.storage[index] > self.storage[parent]:
            self._swap(index, parent)
            # recursively call until the if condition is not met
            self._bubble_up(parent)
for the sift down, it will continue to sift down until: 
        # if the largest is now not the original index, swap and recursively call sift down
        # until the largest is at the index
        if largest != index:
            self._swap(index, largest)
            self._sift_down(largest)

Reply:
I see the recursive calls now.  Good stuff!